### ISSUE TITLE
lldb set breakpoint bug

### ISSUE THREAD

**Issue by wjllz** [llvm/llvm-project#54479]
Hey, I meet a bug(maybe a feature) which in macos 12.3, and my lldb is the newest one.
Like this:
![image](https://user-images.githubusercontent.com/28553622/159277875-2645c1ce-c556-4c57-954d-7bc510d1280d.png)
If I want set a breakpoint at here, the code value will be set to 0xcc(int3).
![image](https://user-images.githubusercontent.com/28553622/159278047-68992b91-3228-4e51-b5af-ac0b0f52c96c.png)
Then the code will be: 
![image](https://user-images.githubusercontent.com/28553622/159278140-d43af073-3370-42ed-9554-e1b554994ce5.png)
Similar mechanism like windbg. But windbg won't change the instruction(in the developer human view)...

I don't test it on windows/linux, I don't know is it a feature or a bug... But I think this will make someone confused, because they will think the edx value from [rsi], obviously, it is from r14...

Sry for my bad English, and best wish!

**Comment by labath** [comment]
I would say this is a bug. I'm not sure what happens on windows, but I do know that lldb-server (used on linux and bsds) deliberately returns the original memory contents (with trap opcodes removed). I guess debugserver (used on macos) is missing that logic (?)

**Comment by wjllz** [comment]
> I would say this is a bug. I'm not sure what happens on windows, but I do know that lldb-server (used on linux and bsds) deliberately returns the original memory contents (with trap opcodes removed). I guess debugserver (used on macos) is missing that logic (?)

I'm not sure it because of debugserver.

I use this to debug macos kernel by two machine debug with firewire. 

like this:

> fwkdp -v
> lldb: kdp-remote localhost

Seems apple implement kdp protocal by theirselves,and I don't know the implemention use debugserver or not(I guess not, because if it used, I think I more should use gdb-remote... sry, I am confused about this...) so should I report the bug(feature) to apple instead of here?

**Comment by labath** [comment]
> > I would say this is a bug. I'm not sure what happens on windows, but I do know that lldb-server (used on linux and bsds) deliberately returns the original memory contents (with trap opcodes removed). I guess debugserver (used on macos) is missing that logic (?)
> 
> I'm not sure it because of debugserver.
> 
> I use this to debug macos kernel by two machine debug with firewire.
> 
> like this:
> 
> > fwkdp -v
> > lldb: kdp-remote localhost

I suppose that depends on how you look at it. Currently, the only logic we have for this is server-side. lldb-server does it here <https://github.com/llvm/llvm-project/blob/main/lldb/source/Host/common/NativeProcessProtocol.cpp#L614>. I guess the debugserver does not have it at all, and the kdp server probably has the same "bug". One can also imagine a client-side implementation of this, which would then presumably work for all targets (servers), though I'm not sure if that's the right way to do it. (In case anyone is interested, the way that gdb gets around this seems to be by unsetting all breakpoints after a stop, and resetting them just before resuming).

> 
> Seems apple implement kdp protocal by theirselves,and I don't know the implemention use debugserver or not(I guess not, because if it used, I think I more should use gdb-remote... sry, I am confused about this...) so should I report the bug(feature) to apple instead of here?

I don't this is a bad place for this bug, since debugserver source if a part of llvm, but I don't know how the apple folks prefer to get their bug reports. I can cc a couple folks though. (@JDevlieghere, @jimingham)

**Comment by wjllz** [comment]
> > > I would say this is a bug. I'm not sure what happens on windows, but I do know that lldb-server (used on linux and bsds) deliberately returns the original memory contents (with trap opcodes removed). I guess debugserver (used on macos) is missing that logic (?)
> > 
> > 
> > I'm not sure it because of debugserver.
> > I use this to debug macos kernel by two machine debug with firewire.
> > like this:
> > > fwkdp -v
> > > lldb: kdp-remote localhost
> 
> I suppose that depends on how you look at it. Currently, the only logic we have for this is server-side. lldb-server does it here https://github.com/llvm/llvm-project/blob/main/lldb/source/Host/common/NativeProcessProtocol.cpp#L614. I guess the debugserver does not have it at all, and the kdp server probably has the same "bug". One can also imagine a client-side implementation of this, which would then presumably work for all targets (servers), though I'm not sure if that's the right way to do it. (In case anyone is interested, the way that gdb gets around this seems to be by unsetting all breakpoints after a stop, and resetting them just before resuming).
> 
> > Seems apple implement kdp protocal by theirselves,and I don't know the implemention use debugserver or not(I guess not, because if it used, I think I more should use gdb-remote... sry, I am confused about this...) so should I report the bug(feature) to apple instead of here?
> 
> I don't this is a bad place for this bug, since debugserver source if a part of llvm, but I don't know how the apple folks prefer to get their bug reports. I can cc a couple folks though. (@JDevlieghere, @jimingham)

thx

**Comment by JDevlieghere** [comment]
> I don't this is a bad place for this bug, since debugserver source if a part of llvm, but I don't know how the apple folks prefer to get their bug reports. I can cc a couple folks though. (@JDevlieghere, @jimingham)

If this is missing from debugserver (@jasonmolenda to confirm) then that's certainly something we can track here. The kernel/KDP case doesn't use debugserver. For that I would recommend filing a bug through https://developer.apple.com/bug-reporting/ so we can track it internally and include other Apple folks that don't necessarily work on lldb/llvm.

**Comment by jasonmolenda** [comment]
debugserver hides the trap instruction, same as lldb-server et al. 

```
(lldb) disass -b -n main
a.out`main:
->  0x100003fa0 <+0>:  55                    pushq  %rbp
    0x100003fa1 <+1>:  48 89 e5              movq   %rsp, %rbp
    0x100003fa4 <+4>:  c7 45 fc 00 00 00 00  movl   $0x0, -0x4(%rbp)
    0x100003fab <+11>: 31 c0                 xorl   %eax, %eax
    0x100003fad <+13>: 5d                    popq   %rbp
    0x100003fae <+14>: c3                    retq   
(lldb) process plugin packet send m100003fa0,f
  packet: m100003fa0,f
response: 554889e5c745fc0000000031c05dc3
(lldb) br s -a 0x100003fa1
Breakpoint 2: where = a.out`main + 1, address = 0x0000000100003fa1
(lldb) process plugin packet send m100003fa0,f
  packet: m100003fa0,f
response: 554889e5c745fc0000000031c05dc3
(lldb) 
```

The KDP stub inside the kernel doesn't seem to do this. Jonas' suggestion of filing a feedback with Apple to request this be handled in the kernel's stub is the correct approach.

**Comment by wjllz** [comment]
Maybe I could write a lldb python script to solve this question:

Like:

``` python
def kernel_set_breakpoint_wrapper(...):
     saved_original_value
     set_a_break_point(wrapper_args)
     rewrite_roginal_value
```

I guess, and I will try🤔

**Comment by labath** [comment]
I doubt that will work. If I correctly understand what you are planning to do, then the rewrite operation will just overwrite the breakpoint causing it to never be hit at all.

**Comment by wjllz** [comment]
> I doubt that will work. If I correctly understand what you are planning to do, then the rewrite operation will just overwrite the breakpoint causing it to never be hit at all.

Yes! I guess. I could just use debugger.HandleCommand to set the breakpoint, and by id to got the location information.

But got original_value is a easy thing, because lldb don't support "memory history" in kdp protocal.  😤

**Comment by wjllz** [comment]
#   my mini write-up

Hey, I solved(maybe) this problem by use lldb python script. 

However, my code is ugly and buggy, but I am not familar with how to implement a debugger, so I decide to make myself be happy -- abandon this. I don't want to update the code, make it more stable and so on... In Chinese, we say:

>   你还要什么自行车🍺

I am sorry about this part... so sry. 

So I decide to write this write-up, If anybody meet same problem as me. maybe this will help them a lot... I hope so...

And sry for my poor English...

>   <font color="green"> I am so sry that I don't read the lldb source code, so something(maybe too much thing....) which I explained maybe is wrong. sry for this part...  </font>

##  The problem

When I want to use lldb to debug macos kernel, which implement kdp protocal to archive this. I found this problem... 

See this code snip:

``` assembly
    0xffffff7f9d69899e: mov    rbx, rax
    0xffffff7f9d6989a1: mov    r13d, 0xe00002c2
    0xffffff7f9d6989a7: test   rbx, rbx
```

if we want to set a breakpoint at here(`|0xffffff7f9d69899e|`). 

>   b 0xffffff7f9d69899e

when we hit the breakpoint, the code snip will like this:

``` assembly
    (lldb) dis -s 0xffffff7f9d69899e
    0xffffff7f9d69899e: int3
    0xffffff7f9d69899f: mov    ebx, eax
    0xffffff7f9d6989a1: mov    r13d, 0xe00002c2
```

It is not a big problem... but makes me not happy... If I forgot this. I will take too much time to think about who is `|ebx|`, who is `|eax|`.

Of course u could use ida to help u solve this, but ida is so expensive to afford it... so I decide to use lldb script to solve this.

##  What happened at here

I don't know why, but I <font color = "green">guess</font> the process like this:

When a debugger want to implement set a breakpoint.

-   U use `|"b 0xffffff7f9d69899e"|` command
-   lldb core solve this command. and try to create a breakpoint. one `|SBBreakPoint|` to log this breakpoint information.
-   lldb start monitor does we hit the breakpoint?
    -   Yes, lldb set the breakpoint address memory be "0xcc(int3)". Then we will stop at here.

Then we hit the breakpoint code, the opcode will like this:

``` c++
0xffffff7f9d69899e 0xcc
```

<font color="green">If we want to use "0xcc" to disassembly this code. it will be wrong...</font>

I should mention again.. I am not sure about this process, so maybe it's wrong...

##  Solve 

Finnaly I use this script to solve this problem:

``` python
"""
[+] macos_kernel_breakpoint_helper.py
[+] 🍺🍺🍺
[+] Description
        =>  (*) To solve macos kernel debug int3 breakpoint
[+] details 
        =>  (*) see https://github.com/llvm/llvm-project/issues/54479
"""

#coding=utf-8
#自定义lldb命令 
from ast import arg
from atexit import register
from ctypes.wintypes import tagRECT
from distutils.log import error
from glob import glob
import string
import lldb
import optparse
import shlex

global_breakpoint_manage_list = []  #   we use a list which's entry is a dictionary to manage the breakpoint information

def cmd_krn_continue(debugger, command, result, __dict):
    global global_breakpoint_manage_list
    
    #   How to get breakpoint by id
    
    error = lldb.SBError()
    target = debugger.GetSelectedTarget()
    
    for dict_entry in global_breakpoint_manage_list:
        if(dict_entry["enabled"] == False): # ugly code, but should work...
            #   recover it
            idx = dict_entry["id"]
            breakpoint_find = target.FindBreakpointByID(idx)
            breakpoint_find.SetEnabled(True)
    debugger.HandleCommand("continue")

def breakpoint_fix_handler(frame, bp_loc, dict):
    global global_breakpoint_manage_list    # global variable, ugly but works
    #   at here, the breakpoint hited...
    
    breakpoint_idx = bp_loc.GetBreakpoint().GetID()    #   we use this to got breakpoint identifier
    
    error = lldb.SBError()
    thread = frame.GetThread()
    process = thread.GetProcess()
    debugger = process.GetTarget().GetDebugger()
    
    map_entry = None
    idx = 0
    for dict_entry in global_breakpoint_manage_list:
        if(dict_entry["id"] == breakpoint_idx):
            map_entry = dict_entry
            global_breakpoint_manage_list[idx]["enabled"] = False   #   set this...
            print("\n👉 : (%d) entry value : %s" %(breakpoint_idx, map_entry))
            break
        idx = idx + 1
    
    #   I should use this, in another word...
    
    #process.WriteMemory(map_entry["address"], bytes( [map_entry["old_value"] ]), error)    # if I didn't set this, what happend?
    bp_loc.GetBreakpoint().SetEnabled(False)    #   Do u think this is helpful?
    
    print("👉 : fix over!!!")

def cmd_krn_set_breakpoint(debugger, command, result, __dict):
    global global_breakpoint_manage_list    # global variable, ugly but works
    
    error = lldb.SBError()
    target = debugger.GetSelectedTarget()

    debugger.HandleCommand("breakpoint " + command) # use this to set the breakpoint
    
    new_breakpoint = list(target.breakpoint_iter())[-1] #   the breakpoint is newest one, so we use the iterator to got it(the last one)
    
    breakpoint_address = list(new_breakpoint)[0].GetLoadAddress() #   the breakpoint set where(address)? ugly but work
    breakpoint_id = new_breakpoint.GetID()
    old_saved_value = target.GetProcess().ReadUnsignedFromMemory(breakpoint_address, 1, error)
    breakpoint_enabled = True
    print("👉 : %s saved old_value: %s" % (hex(breakpoint_address), hex(old_saved_value)))
    dict_entry = {"address" : breakpoint_address, "id" : breakpoint_id, "old_value" : old_saved_value, "enabled" : breakpoint_enabled}

    
    #   now, we saved all infromatioin which we need...
    global_breakpoint_manage_list.append(dict_entry)   
    
    #   set a breakpoint hook
    new_breakpoint.SetScriptCallbackFunction("macos_kernel_breakpoint_helper.breakpoint_fix_handler")
    print("👉 : set a breakpoint handler success, going!")

def __lldb_init_module(debugger, internal_dict):  
    debugger.HandleCommand('command script add -f macos_kernel_breakpoint_helper.cmd_krn_set_breakpoint krn_set_breakpoint')  
    debugger.HandleCommand('command script add -f macos_kernel_breakpoint_helper.cmd_krn_continue krn_continue')  
    print("👉 : our script inited!")
```

This is my examine result:


![image](https://user-images.githubusercontent.com/28553622/162032151-bc40b512-f8f2-4f82-9a56-04b475728bf6.png)

You could see it works.

##  What happened(the details)

I create two command.

- [x] `|krn_set_breakpoint|`
- [x] `|krn_continue|`

U could guess their usage by their name.

The code is buggy and ugly, I originally have a plan , but I found it failed... So I change it to use plan B. fortunately, it works.

But I don't want to remove the code which is of Plan A, becuase I think it will help u understand plan B.

brief summary:

- [x] plan A
    -   change the address memory.
- [x] plan B
    -   change the breakpoint state

### Plan A

If u remember what I said above, u maybe remembre that I say it is because of lldb change the address memory to "0xcc" when a breakpoiint hit.

So my original plan is:

>   <font color ="red">If we overwrite the address memory to original value instead of "0xcc" when the breakpoint is hit.</font> maybe everything is ok.

lldb use SBBreakpoint to manage breakpoint which u set. If u read the document. u will found function `|SetScriptCallbackFunction|`. use this function u could set a callback function which will be trigger when your breakpoint be triggerd:

``` python
    #   set a breakpoint hook
    new_breakpoint.SetScriptCallbackFunction("macos_kernel_breakpoint_helper.breakpoint_fix_handler")
    print("👉 : set a breakpoint handler success, going!")
```

So u just need to do something in `|breakpoint_fix_handler|`, like just overwrite the orignal value to "0xcc":

``` python
def breakpoint_fix_handler(frame, bp_loc, dict):
    global global_breakpoint_manage_list    # global variable, ugly but works
    #   at here, the breakpoint hited...
    
    breakpoint_idx = bp_loc.GetBreakpoint().GetID()    #   we use this to got breakpoint identifier
    
    error = lldb.SBError()
    thread = frame.GetThread()
    process = thread.GetProcess()
    debugger = process.GetTarget().GetDebugger()
    
    map_entry = None
    for dict_entry in global_breakpoint_manage_list:
        if(dict_entry["id"] == breakpoint_idx):
            map_entry = dict_entry
            print("\n👉 : (%d) entry value : %s" %(breakpoint_idx, map_entry))
            break
    
    process.WriteMemory(map_entry["address"], bytes( [map_entry["old_value"] ]), error)    # if I didn't set this, what happend?
    print("👉 : fix over!!!")

```

There are tow question:

-   [x] it is works?
-   [x] how to got the original value of the address memory

The first question seems is easy to understand. "No". But I want to explain why long long time ago...

Now, let's try to answer the second question, how to get the original(old) value of the adderss memory.

As I mentioined, lldb core use SBBreakpoiint to manage the breakpoint which we set, and u could use a `|iterator|` to got all breakpoints:

``` python
    new_breakpoint = list(target.breakpoint_iter())[-1] #   the breakpoint is newest one, so we use the iterator to got it(the last one)
```

Maybe u are curious about what does `|-1|` means?

we use this command to set a new breakpoint, so the breakpoint is the `|last one|`, so we use `|-1|` to got it.

Then a SBBreakpoint saved some infomration about it:

-   id: to distinct it with others
-   address: set at where(I use "address" mention this above)
-   another: like condition....

To got the address I don't find any useful API, so I use this:

``` python
    breakpoint_address = list(new_breakpoint)[0].GetLoadAddress() #   the breakpoint set where(address)? ugly but work
```

As I mentioined before, <font color = "green">Only If the breakpoint been hited, it will be setted to "0xcc", it means, when we just set the breakpoint, and the breakpoint hasn't been hitted, it will be orinal value instead of "0xcc"</font>

So we just read it!

``` python
    old_saved_value = target.GetProcess().ReadUnsignedFromMemory(breakpoint_address, 1, error)
```

And I said that SBBreakpoint has "id", make a dictionay list, and use "id" to distinct them:

``` python
    breakpoint_address = list(new_breakpoint)[0].GetLoadAddress() #   the breakpoint set where(address)? ugly but work
    breakpoint_id = new_breakpoint.GetID()
    old_saved_value = target.GetProcess().ReadUnsignedFromMemory(breakpoint_address, 1, error)
    breakpoint_enabled = True
    print("👉 : %s saved old_value: %s" % (hex(breakpoint_address), hex(old_saved_value)))
    dict_entry = {"address" : breakpoint_address, "id" : breakpoint_id, "old_value" : old_saved_value, "enabled" : breakpoint_enabled}
    
    global_breakpoint_manage_list.append(dict_entry)   
```

The we just use `|id|` to decide which memory should be written as which value..

This is plan A try to do.

However, I found it can't worked...

####    What's wrong in plan A

>   Note: this part I guess I am undersatnd wrongly, I just guess what happend, so sry about this part...

In plan A, we try to overwrite the address memory to original value instead of "0XCC", when we hit the breakpoint, we could see:

``` c++
  ->0xffffff7f9d69899e: mov    rbx, rax
    0xffffff7f9d6989a1: mov    r13d, 0xe00002c2
    0xffffff7f9d6989a7: test   rbx, rbx
```

Seems it works, But If u use si to execute next instruction, u will found...

``` c++
    0xffffff7f9d69899e: int3
    0xffffff7f9d69899f: mov    ebx, eax
  ->0xffffff7f9d6989a1: mov    r13d, 0xe00002c2
```

Boom... So what happend?

<font color = "green">I guess that:</font>

-   When a breakpoint been hit, it will be okey becuase of that we have overwrite the memory to orinal value.
-   When we execute next command, Think about how to implent a step-in in a debugger
    -   I guess for each instruction, lldb create a context. I mean, If u use "si" to execute different instruction, the context will be created(or be update?).
    -   When we hit the breakpoint, now we are in context A
    -   When we execute next instructioin, now we are in context B
        -   lldb found we change to another context, it will update related information
        -   lldb core scan breakpoint, found breakpoint which has been setted by us. found the value is not "0xcc", but the hit count is not 0(I don't know this process need now hit count or not? I just think maybe need)... then lldb core use 0xcc to overwrite the address memory content...
-   Boom part, now the error occurs again...

So I change it to plan B

### Plan B

If u understand plan A, plan B will be easy... I will explain it less details...

I think overwrite address memory won't work... fortunately, I found this API:

``` python
    bp_loc.GetBreakpoint().SetEnabled(False)    #   Do u think this is helpful?
```

If u set it to Flase, the breakpoint will be disabled... Then everything is ok!
the context been updated, lldb core will help us recover the value...

>   note: I guess lldb core use a way to manage adress moery orinal value... But in SBBreakpoint I don't find any API about it, I guess it use "context" to manage it?

Then when we continue, we just reset it to True again:

``` python
def cmd_krn_continue(debugger, command, result, __dict):
    global global_breakpoint_manage_list
    
    #   How to get breakpoint by id
    
    error = lldb.SBError()
    target = debugger.GetSelectedTarget()
    
    for dict_entry in global_breakpoint_manage_list:
        if(dict_entry["enabled"] == False): # ugly code, but should work...
            #   recover it
            idx = dict_entry["id"]
            breakpoint_find = target.FindBreakpointByID(idx)
            breakpoint_find.SetEnabled(True)
    debugger.HandleCommand("continue")
```


##  Conclusion

Too much bug in my code... sry again, but I hope u will enjoy this... sry for my bad English again....

**Comment by jimingham** [comment]
Both debugserver and the KDP process stub send breakpoint setting packets, they don't modify target memory directly.  Stubs that implement a breakpoint setting packet have to return the actual memory at the breakpoint, since lldb has no way of knowing how the breakpoint was implemented and wouldn't a priori know how to remove it.  I just checked, debugserver certainly does this and though it's been a while since I did KDP debugging, last time I looked it worked that way as well.

Jim
